File: product-of-array-except-self/yeonguchoe.java
class Solution {
    // Time complexity: O(n)
    // Space complexity: O(n)
    public int[] productExceptSelf(int[] nums) {
        int[] leftMultiplier = new int[nums.length];
        int[] rightMultiplier = new int[nums.length];

        leftMultiplier[0] = 1;
        for (int i = 0; i < nums.length - 1; i++) {
            leftMultiplier[i + 1] = nums[i] * leftMultiplier[i];
        }

        rightMultiplier[nums.length - 1] = 1;
        for (int i = nums.length - 1; i > 0; i--) {
            rightMultiplier[i - 1] = nums[i] * rightMultiplier[i];
        }

        int[] result = new int[nums.length];
        for (int i = 0; i < result.length; i++) {
            result[i] = leftMultiplier[i] * rightMultiplier[i];
        }
        return result;
    }
}

File: maximum-product-subarray/yeonguchoe.cpp
class Solution {
    // Time complexity: O(n)
    // Space complexity: O(1)
public:
    int maxProduct(vector<int>& nums) {

        int result = nums[0];
        int previous_max = 1;
        int previous_min = 1;

        for (int num : nums) {
            int temp = previous_max;
            previous_max = max({num, previous_max * num, previous_min * num});
            previous_min = min({num, temp * num, previous_min * num});
            result = max(result, previous_max);
        }

        return result;
    }
};
